combinetf2/fitter.py

```diff
@@ -539,7 +541,132 @@ def global_impacts_parms(self):
+            )
+        else:
+            # pd2ldbeta2 is diagonal, so we can use gradient instead of jacobian
+            pd2ldbeta2_diag = t2.gradient(pdldbeta, self.ubeta)
+            pd2ldbeta2 = tf.linalg.LinearOperatorDiag(
+                pd2ldbeta2_diag, is_self_adjoint=True
+            )
+        return pd2ldbeta2
+
+    def _dxdvars(self):
+        with tf.GradientTape() as t2:
+            t2.watch([self.theta0, self.nobs, self.beta0])
+            with tf.GradientTape() as t1:
+                t1.watch([self.theta0, self.nobs, self.beta0])
+                val = self._compute_loss()
+            grad = t1.gradient(val, self.x)
+        pd2ldxdtheta0, pd2ldxdnobs, pd2ldxdbeta0 = t2.jacobian(
+            grad, [self.theta0, self.nobs, self.beta0], unconnected_gradients="zero"
+        )
+
+        # cov is inverse hesse, thus cov ~ d2xd2l
+        dxdtheta0 = -self.cov @ pd2ldxdtheta0
+        dxdnobs = -self.cov @ pd2ldxdnobs
+        dxdbeta0 = -self.cov @ pd2ldxdbeta0
+
+        return dxdtheta0, dxdnobs, dxdbeta0
+
+    def _chi2(
+        self,
+        fun,
+        ndf_reduction=0,
+        profile=False,
+    ):
+
+        with tf.GradientTape() as t:
+            t.watch([self.theta0, self.nobs, self.beta0])
+            expected = self._compute_expected(
+                fun,
+                inclusive=True,
+                profile=profile,
+                full=False,
+                need_observables=True,
+            )
+            observed = fun(None, self.nobs)
+            residuals = expected - observed
+
+            residuals_flat = tf.reshape(residuals, (-1,))
+        pdresdx, pdresdtheta0, pdresdnobs, pdresdbeta0 = t.jacobian(
+            residuals_flat,
+            [self.x, self.theta0, self.nobs, self.beta0],
+            unconnected_gradients="zero",
+        )
+
+        if profile:
+            # apply chain rule to take into account correlations with the fit parameters
+            dxdtheta0, dxdnobs, dxdbeta0 = self._dxdvars()
+
+            dresdtheta0 = pdresdtheta0 + pdresdx @ dxdtheta0
+            dresdnobs = pdresdnobs + pdresdx @ dxdnobs
+            dresdbeta0 = pdresdbeta0 + pdresdx @ dxdbeta0
+
+            var_theta0 = tf.where(
+                self.indata.constraintweights == 0.0,
+                tf.zeros_like(self.indata.constraintweights),
+                tf.math.reciprocal(self.indata.constraintweights),
+            )
+
+            res_cov = dresdtheta0 @ (var_theta0[:, None] * tf.transpose(dresdtheta0))
+        else:
+            # full dependency in partial derivatives
+            dresdnobs = pdresdnobs
+            dresdbeta0 = pdresdbeta0
+
+            res_cov = pdresdx @ tf.matmul(self.cov, pdresdx, transpose_b=True)
+
+        res_cov_stat = dresdnobs @ (self.nobs[:, None] * tf.transpose(dresdnobs))
+        res_cov += res_cov_stat
+
+        if self.binByBinStat:
+            var_beta0 = tf.math.reciprocal(self.kstat)
```

Comment:
This should be 1.0 for binByBinStatType normal, and beta0/kstat for gamma
(the proper calculation of this could also be automated based on the appropriate gradients, in general it's (d2lbeta/dbeta2)/(d2lbeta/dbetadbeta0)^2 )

Some special handling is also needed for the betamask case for gamma uncertainties (where it would be sufficient to set var_beta0 to 0)